tests/kafkatest/tests/client/downconversion_memory_test.py

```diff
@@ -0,0 +1,113 @@
+        self.message_size = 1024
+        self.batch_size = self.message_size * 50
+        self.num_partitions = 200
+        self.max_fetch_size = 200 * 1024 * 1024
+        self.num_producers = 1
+        self.num_consumers = 1
+        self.topics = ["test_topic"]
+        self.zk = ZookeeperService(self.test_context, num_nodes=1)
+        self.heap_dump_path = "/mnt/heapdump/"
+        self.consumer_version = LATEST_0_10
+
+    def setUp(self):
+        self.zk.start()
+
+    def tearDown(self):
+        for node in self.kafka.nodes:
+            node.account.ssh("rm -rf %s" % self.heap_dump_path, allow_fail=False)
+
+    @cluster(num_nodes=12)
+    @parametrize(version=str(LATEST_1_1))
+    @parametrize(version=str(DEV_BRANCH))
+    def test_downconversion(self, version):
+        expect_out_of_memory = False
+        out_of_memory = False
+        if KafkaVersion(version) <= LATEST_1_1:
+            expect_out_of_memory = True
+
+        self.kafka = KafkaService(self.test_context, num_nodes=1, zk=self.zk, version=KafkaVersion(version),
+                                  topics={topic:
+                                              {"partitions": self.num_partitions,
+                                               "replication-factor": 1,
+                                               "configs": {"min.insync.replicas": 1}}
+                                          for topic in self.topics},
+                                  heap_opts="-Xmx%dM -Xms%dM -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath=%s" % (self.heap_size, self.heap_size, self.heap_dump_path),
+                                  do_logging=False)
+        for node in self.kafka.nodes:
+            node.account.ssh("mkdir -p %s" % self.heap_dump_path, allow_fail=False)
+        self.kafka.start()
+
+        # seed kafka with messages
+        for topic in self.topics:
+            producer = ProducerPerformanceService(
+                self.test_context, self.num_producers, self.kafka, topic=topic,
+                num_records=self.max_messages, record_size=self.message_size, throughput=-1,
+                version=KafkaVersion(version),
+                settings={
+                    'acks': 1,
+                    'batch.size': self.batch_size
+                }
+            )
+            producer.run()
+            print "Producer throughput: "
+            print(compute_aggregate_throughput(producer))
+
+        # consume
+        for topic in self.topics:
+            consumer = ConsumerPerformanceService(self.test_context, self.num_consumers, self.kafka, topic=topic,
+                                                  messages=self.max_messages, version=self.consumer_version,
+                                                  config={"fetch.max.bytes": self.max_fetch_size}, timeout=60)
+            consumer.run()
+
+        for node in self.kafka.nodes:
+            if self.kafka.file_exists(node, self.heap_dump_path + "*.hprof"):
```

Comment:
If we run out of memory, I am wondering if we may go into the state where we cannot consume "max_messages" for example and test will fail with timeout error on consumer.run(). I understand that  run() does start(), wait(timeout = 10 mins), stop(), where wait will wait for given number of messages to be consumed. Maybe I am missing something, but wanted to check.

Reply:
Good question. I added a timeout to ConsumerPerformanceService#_worker to make this work. The idea being that if we run out of memory, we may never be able to consume messages but will timeout in 60 seconds. This causes us to return "normally" from the wait() call.